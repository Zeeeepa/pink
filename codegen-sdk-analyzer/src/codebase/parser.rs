use codegen_sdk_ast::{Definitions, References, input::File};
#[cfg(feature = "serialization")]
use codegen_sdk_common::serialize::Cache;
use indicatif::{ProgressBar, ProgressStyle};

use super::discovery::{FilesToParse, log_languages};
use crate::{
    ParsedFile,
    database::{CodegenDatabase, Db},
    parser::parse_file,
};
fn execute_op_with_progress<Database: Db + ?Sized + 'static, T: Send + Sync>(
    db: &Database,
    files: FilesToParse,
    name: &str,
    op: fn(&Database, File) -> T,
) -> Vec<T> {
    let multi = db.multi_progress();
    let style = ProgressStyle::with_template(
        "[{elapsed_precise}] {wide_bar} {msg} [{per_sec}] [estimated time remaining: {eta}]",
    )
    .unwrap();
    let pg = multi.add(
        ProgressBar::new(files.files(db).len() as u64)
            .with_style(style)
            .with_message(name.to_string()),
    );
    let inputs = files
        .files(db)
        .into_iter()
        .map(|file| (&pg, file, op))
        .collect::<Vec<_>>();
    let results: Vec<T> = salsa::par_map(db, inputs, move |db, input| {
        let (pg, file, op) = input;
        let res = op(
            db,
            #[cfg(feature = "serialization")]
            &cache,
            file,
        );
        pg.inc(1);
        res
    });
    pg.finish();
    multi.remove(&pg);
    results
}
#[salsa::tracked]
fn parse_files_par(db: &dyn Db, files: FilesToParse) {
    let _: Vec<_> = execute_op_with_progress(db, files, "Parsing Files", |db, file| {
        parse_file(db, file);
    });
}
#[salsa::tracked]
fn parse_files_definitions_par(db: &dyn Db, files: FilesToParse) {
    let _: Vec<_> = execute_op_with_progress(db, files, "Parsing Definitions", |db, file| {
        let file = parse_file(db, file);
        if let Some(parsed) = file.file(db) {
            #[cfg(feature = "typescript")]
            if let ParsedFile::Typescript(parsed) = parsed {
                parsed.definitions(db);
                parsed.references(db);
            }
            #[cfg(feature = "python")]
            if let ParsedFile::Python(parsed) = parsed {
                parsed.definitions(db);
                parsed.references(db);
            }
        }
        ()
    });
}
pub fn parse_files<'db>(
    db: &'db CodegenDatabase,
    #[cfg(feature = "serialization")] cache: &'db Cache,
    files_to_parse: FilesToParse,
) -> () {
    rayon::ThreadPoolBuilder::new()
        .stack_size(1024 * 1024 * 1024 * 10)
        .build_global()
        .unwrap();
    log_languages();
    #[cfg(feature = "serialization")]
    let cache = Cache::new().unwrap();
    #[cfg(feature = "serialization")]
    let cached = get_cached_count(&cache, &files_to_parse);
    log::info!("Parsing {} files", files_to_parse.files(db).len());
    parse_files_par(
        db,
        #[cfg(feature = "serialization")]
        &cache,
        files_to_parse,
    );
    log::info!("Parsing definitions");
    parse_files_definitions_par(
        db,
        #[cfg(feature = "serialization")]
        &cache,
        files_to_parse,
    );
    #[cfg(feature = "serialization")]
    report_cached_count(cached, &files_to_parse.files(db));
}
