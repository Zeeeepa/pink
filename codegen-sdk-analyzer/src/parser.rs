use codegen_sdk_common::FileNodeId;
use codegen_sdk_cst::CSTLanguage;
use codegen_sdk_macros::{languages_ast, parse_language};
languages_ast!();

#[salsa::tracked(return_ref)]
pub fn parse_file<'db>(
    db: &'db dyn codegen_sdk_resolution::Db,
    file: codegen_sdk_common::FileNodeId,
) -> Option<ParsedFile<'db>> {
    parse_language!();
    None
}
