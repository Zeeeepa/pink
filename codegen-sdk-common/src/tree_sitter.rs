use tree_sitter::TextProvider;

pub fn get_text_from_node(node: &tree_sitter::Node, source: &str) -> String {
    String::from_utf8(source.as_bytes().text(*node).next().unwrap().to_vec()).unwrap()
}
