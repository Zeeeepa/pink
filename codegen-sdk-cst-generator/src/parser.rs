use std::error::Error;

use serde::{Deserialize, Serialize};

#[derive(Debug, Serialize, Deserialize)]
pub struct Node {
    #[serde(rename = "type")]
    pub type_name: String,
    pub named: bool,
    #[serde(default)]
    pub subtypes: Vec<TypeDefinition>,
    #[serde(default)]
    pub fields: Option<Fields>,
    #[serde(default)]
    pub children: Option<Children>,
}

#[derive(Debug, Serialize, Deserialize)]
pub struct Fields {
    #[serde(flatten)]
    pub fields: std::collections::HashMap<String, FieldDefinition>,
}

#[derive(Debug, Serialize, Deserialize)]
pub struct FieldDefinition {
    pub multiple: bool,
    pub required: bool,
    #[serde(default)]
    pub types: Vec<TypeDefinition>,
}

#[derive(Debug, Serialize, Deserialize, Clone)]
pub struct TypeDefinition {
    #[serde(rename = "type")]
    pub type_name: String,
    pub named: bool,
}

#[derive(Debug, Serialize, Deserialize)]
pub struct Children {
    pub multiple: bool,
    pub required: bool,
    #[serde(default)]
    pub types: Vec<TypeDefinition>,
}

pub fn parse_node_types(source: &str) -> Result<Vec<Node>, Box<dyn Error>> {
    let parsed: Vec<Node> = serde_json::from_str(&source)?;
    Ok(parsed)
}
#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_parse_node_types() {
        let source = tree_sitter_python::NODE_TYPES;
        let cst = parse_node_types(source).unwrap();
        panic!();
    }
}
