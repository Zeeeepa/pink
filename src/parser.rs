use std::path;

use codegen_sdk_analyzer::{CodegenDatabase, Db, Parsed};
#[cfg(feature = "serialization")]
use codegen_sdk_common::serialize::Cache;
use indicatif::ProgressBar;

use crate::discovery::{FilesToParse, collect_files, log_languages};
fn parse_file<'db>(
    db: &'db dyn Db,
    #[cfg(feature = "serialization")] cache: &Cache,
    file: &path::PathBuf,
) -> Parsed<'db> {
    if file.is_dir() {
        log::warn!("Skipping directory: {}", file.display());
        return Parsed::new(db, None);
    }
    let result = db.input(file.into());
    return match result {
        Ok(program) => codegen_sdk_analyzer::parse_file(db, program),
        Err(e) => {
            log::error!("Error parsing file {}: {}", file.display(), e);
            Parsed::new(db, None)
        }
    };
}
#[salsa::tracked]
fn parse_files_par(db: &dyn Db, files: FilesToParse) {
    let multi = db.multi_progress();
    let pg = multi.add(ProgressBar::new(files.files(db).len() as u64));
    let inputs = files
        .files(db)
        .into_iter()
        .map(|file| (&pg, file))
        .collect::<Vec<_>>();
    let _: Vec<()> = salsa::par_map(db, inputs, |db, input| {
        let (pg, file) = input;
        parse_file(
            db,
            #[cfg(feature = "serialization")]
            &cache,
            &file,
        );
        pg.inc(1);
        ()
    });
    pg.finish();
    multi.remove(&pg);
}
pub fn parse_files<'db>(
    db: &'db CodegenDatabase,
    #[cfg(feature = "serialization")] cache: &'db Cache,
    dir: String,
) -> (FilesToParse, Vec<String>) {
    rayon::ThreadPoolBuilder::new()
        .stack_size(1024 * 1024 * 1024 * 10)
        .build_global()
        .unwrap();
    let (tx, rx) = crossbeam::channel::unbounded();
    let mut errors = Vec::new();
    log_languages();
    #[cfg(feature = "serialization")]
    let cache = Cache::new().unwrap();
    #[cfg(feature = "serialization")]
    let cached = get_cached_count(&cache, &files_to_parse);
    let files_to_parse = collect_files(db, dir);
    log::info!("Parsing {} files", files_to_parse.files(db).len());
    parse_files_par(
        db,
        #[cfg(feature = "serialization")]
        &cache,
        files_to_parse,
    );
    drop(tx);
    #[cfg(feature = "serialization")]
    report_cached_count(cached, &files_to_parse.files(db));
    for e in rx.iter() {
        errors.push(e);
    }
    (files_to_parse, errors)
}
