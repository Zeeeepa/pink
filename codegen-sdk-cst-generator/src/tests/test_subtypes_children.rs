use assert_tokenstreams_eq::assert_tokenstreams_eq;
use codegen_sdk_common::parser::{Children, Node, TypeDefinition};
use quote::quote;

use crate::{generate_cst, test_util::get_language};

#[test_log::test]
fn test_subtypes_with_children() {
    let nodes = vec![
        // A block can contain multiple statements
        Node {
            type_name: "block".to_string(),
            subtypes: vec![],
            named: true,
            root: false,
            fields: None,
            children: Some(Children {
                multiple: true,
                required: false,
                types: vec![TypeDefinition {
                    type_name: "statement".to_string(),
                    named: true,
                }],
            }),
        },
        // Statement is a subtype with its own subtypes
        Node {
            type_name: "statement".to_string(),
            subtypes: vec![
                TypeDefinition {
                    type_name: "if_statement".to_string(),
                    named: true,
                },
                TypeDefinition {
                    type_name: "return_statement".to_string(),
                    named: true,
                },
            ],
            named: true,
            root: false,
            fields: None,
            children: None,
        },
        // Concrete statement types
        Node {
            type_name: "if_statement".to_string(),
            subtypes: vec![],
            named: true,
            root: false,
            fields: None,
            children: Some(Children {
                multiple: false,
                required: true,
                types: vec![TypeDefinition {
                    type_name: "block".to_string(),
                    named: true,
                }],
            }),
        },
        Node {
            type_name: "return_statement".to_string(),
            subtypes: vec![],
            named: true,
            root: false,
            fields: None,
            children: None,
        },
    ];
    let language = get_language(nodes);
    let output = generate_cst(&language).unwrap();
    let expected = quote! {
        use bytes::Bytes;
        use codegen_sdk_common::*;
        use derive_more::Debug;
        use rkyv::{Archive, Deserialize, Serialize};
        use subenum::subenum;
        use tree_sitter;

        #[derive(Debug, Clone)]
        pub struct Block {
            start_byte: usize,
            end_byte: usize,
            _kind: String,
            #[debug("[{},{}]", start_position.row, start_position.column)]
            start_position: Point,
            #[debug("[{},{}]", end_position.row, end_position.column)]
            end_position: Point,
            #[debug(ignore)]
            buffer: Arc<Bytes>,
            #[debug(ignore)]
            kind_id: u16,
            children: Vec<Statement>,
        }

        impl HasChildren for Block {
            type Child = Statement;
            fn children(&self) -> Vec<Self::Child> {
                self.children.iter().cloned().collect()
            }
            fn children_by_field_name(&self, field_name: &str) -> Vec<Self::Child> {
                match field_name {
                    _ => vec![],
                }
            }
        }
    };
    assert_tokenstreams_eq!(&expected, &output);
}
