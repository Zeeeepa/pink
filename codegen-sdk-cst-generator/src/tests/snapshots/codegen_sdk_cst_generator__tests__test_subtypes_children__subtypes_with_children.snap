---
source: codegen-sdk-cst-generator/src/tests/test_subtypes_children.rs
expression: "crate::test_util::snapshot_string(&output)"
---
use std::sync::Arc;
use tree_sitter;
use codegen_sdk_common::*;
use subenum::subenum;
use std::backtrace::Backtrace;
use bytes::Bytes;
use derive_generic_visitor::Drive;
use ambassador::Delegate;
use codegen_sdk_cst::CSTLanguage;
use codegen_sdk_common::language::Language;
#[subenum(BlockChildren, IfStatementChildren, Statement)]
#[derive(Debug, Clone, Eq, PartialEq, Drive, Hash, salsa::Update, Delegate)]
#[delegate(CSTNode<'db1>)]
pub enum NodeTypes<'db1> {
    #[subenum(IfStatementChildren)]
    Block(Block<'db1>),
    #[subenum(Statement, BlockChildren)]
    IfStatement(IfStatement<'db1>),
    #[subenum(Statement, BlockChildren)]
    ReturnStatement(ReturnStatement<'db1>),
}
impl<'db3> From<Block<'db3>> for NodeTypes<'db3> {
    fn from(node: Block<'db3>) -> Self {
        Self::Block(node)
    }
}
impl<'db3> From<IfStatement<'db3>> for NodeTypes<'db3> {
    fn from(node: IfStatement<'db3>) -> Self {
        Self::IfStatement(node)
    }
}
impl<'db3> From<ReturnStatement<'db3>> for NodeTypes<'db3> {
    fn from(node: ReturnStatement<'db3>) -> Self {
        Self::ReturnStatement(node)
    }
}
impl<'db4> FromNode<'db4> for BlockChildren<'db4> {
    fn from_node(
        db: &'db4 dyn salsa::Database,
        node: tree_sitter::Node,
        buffer: &Arc<Bytes>,
    ) -> Result<Self, ParseError> {
        match node.kind_id() {
            0u16 => {
                Ok(Self::ReturnStatement(ReturnStatement::from_node(db, node, buffer)?))
            }
            _ => {
                Err(ParseError::UnexpectedNode {
                    node_type: node.kind().to_string(),
                    backtrace: Backtrace::capture(),
                })
            }
        }
    }
}
impl<'db4> FromNode<'db4> for IfStatementChildren<'db4> {
    fn from_node(
        db: &'db4 dyn salsa::Database,
        node: tree_sitter::Node,
        buffer: &Arc<Bytes>,
    ) -> Result<Self, ParseError> {
        match node.kind_id() {
            0u16 => Ok(Self::Block(Block::from_node(db, node, buffer)?)),
            _ => {
                Err(ParseError::UnexpectedNode {
                    node_type: node.kind().to_string(),
                    backtrace: Backtrace::capture(),
                })
            }
        }
    }
}
impl<'db4> FromNode<'db4> for Statement<'db4> {
    fn from_node(
        db: &'db4 dyn salsa::Database,
        node: tree_sitter::Node,
        buffer: &Arc<Bytes>,
    ) -> Result<Self, ParseError> {
        match node.kind_id() {
            0u16 => {
                Ok(Self::ReturnStatement(ReturnStatement::from_node(db, node, buffer)?))
            }
            _ => {
                Err(ParseError::UnexpectedNode {
                    node_type: node.kind().to_string(),
                    backtrace: Backtrace::capture(),
                })
            }
        }
    }
}
#[derive(Debug, Clone, Drive, Eq, PartialEq, salsa::Update)]
pub struct Block<'db> {
    #[drive(skip)]
    start_byte: usize,
    #[drive(skip)]
    end_byte: usize,
    #[drive(skip)]
    _kind: std::string::String,
    #[drive(skip)]
    start_position: Point<'db>,
    #[drive(skip)]
    end_position: Point<'db>,
    #[drive(skip)]
    buffer: Arc<Bytes>,
    #[drive(skip)]
    kind_id: u16,
    #[drive(skip)]
    is_error: bool,
    #[drive(skip)]
    named: bool,
    #[drive(skip)]
    id: usize,
    pub _children: Vec<BlockChildren<'db>>,
}
impl<'db> FromNode<'db> for Block<'db> {
    fn from_node(
        db: &'db dyn salsa::Database,
        node: tree_sitter::Node,
        buffer: &Arc<Bytes>,
    ) -> Result<Self, ParseError> {
        let start_position = Point::from(db, node.start_position());
        let end_position = Point::from(db, node.end_position());
        Ok(Self {
            start_byte: node.start_byte(),
            end_byte: node.end_byte(),
            _kind: node.kind().to_string(),
            start_position: start_position,
            end_position: end_position,
            buffer: buffer.clone(),
            kind_id: node.kind_id(),
            is_error: node.is_error(),
            named: node.is_named(),
            id: node.id(),
            _children: named_children_without_field_names(db, node, buffer)?,
        })
    }
}
impl<'db> CSTNode<'db> for Block<'db> {
    fn kind(&self) -> &str {
        &self._kind
    }
    fn start_byte(&self) -> usize {
        self.start_byte
    }
    fn end_byte(&self) -> usize {
        self.end_byte
    }
    fn start_position(&self) -> Point<'db> {
        self.start_position
    }
    fn end_position(&self) -> Point<'db> {
        self.end_position
    }
    fn buffer(&self) -> &Bytes {
        &self.buffer
    }
    fn kind_id(&self) -> u16 {
        self.kind_id
    }
    fn is_error(&self) -> bool {
        self.is_error
    }
    fn is_named(&self) -> bool {
        self.named
    }
    fn id(&self) -> usize {
        self.id
    }
}
impl<'db1> HasChildren<'db1> for Block<'db1> {
    type Child = BlockChildren<'db1>;
    fn children(&self) -> Vec<Self::Child> {
        let mut children: Vec<_> = self._children.iter().cloned().collect();
        children.sort_by_key(|c| c.start_byte());
        children
    }
    fn children_by_field_name(&self, field_name: &str) -> Vec<Self::Child> {
        match field_name {
            _ => vec![],
        }
    }
    fn children_by_field_id(&self, field_id: u16) -> Vec<Self::Child> {
        match field_id {
            _ => vec![],
        }
    }
}
impl<'db> std::hash::Hash for Block<'db> {
    fn hash<H: std::hash::Hasher>(&self, state: &mut H) {
        self.id.hash(state);
    }
}
#[derive(Debug, Clone, Drive, Eq, PartialEq, salsa::Update)]
pub struct IfStatement<'db> {
    #[drive(skip)]
    start_byte: usize,
    #[drive(skip)]
    end_byte: usize,
    #[drive(skip)]
    _kind: std::string::String,
    #[drive(skip)]
    start_position: Point<'db>,
    #[drive(skip)]
    end_position: Point<'db>,
    #[drive(skip)]
    buffer: Arc<Bytes>,
    #[drive(skip)]
    kind_id: u16,
    #[drive(skip)]
    is_error: bool,
    #[drive(skip)]
    named: bool,
    #[drive(skip)]
    id: usize,
    pub _children: Vec<IfStatementChildren<'db>>,
}
impl<'db> FromNode<'db> for IfStatement<'db> {
    fn from_node(
        db: &'db dyn salsa::Database,
        node: tree_sitter::Node,
        buffer: &Arc<Bytes>,
    ) -> Result<Self, ParseError> {
        let start_position = Point::from(db, node.start_position());
        let end_position = Point::from(db, node.end_position());
        Ok(Self {
            start_byte: node.start_byte(),
            end_byte: node.end_byte(),
            _kind: node.kind().to_string(),
            start_position: start_position,
            end_position: end_position,
            buffer: buffer.clone(),
            kind_id: node.kind_id(),
            is_error: node.is_error(),
            named: node.is_named(),
            id: node.id(),
            _children: named_children_without_field_names(db, node, buffer)?,
        })
    }
}
impl<'db> CSTNode<'db> for IfStatement<'db> {
    fn kind(&self) -> &str {
        &self._kind
    }
    fn start_byte(&self) -> usize {
        self.start_byte
    }
    fn end_byte(&self) -> usize {
        self.end_byte
    }
    fn start_position(&self) -> Point<'db> {
        self.start_position
    }
    fn end_position(&self) -> Point<'db> {
        self.end_position
    }
    fn buffer(&self) -> &Bytes {
        &self.buffer
    }
    fn kind_id(&self) -> u16 {
        self.kind_id
    }
    fn is_error(&self) -> bool {
        self.is_error
    }
    fn is_named(&self) -> bool {
        self.named
    }
    fn id(&self) -> usize {
        self.id
    }
}
impl<'db1> HasChildren<'db1> for IfStatement<'db1> {
    type Child = IfStatementChildren<'db1>;
    fn children(&self) -> Vec<Self::Child> {
        let mut children: Vec<_> = self._children.iter().cloned().collect();
        children.sort_by_key(|c| c.start_byte());
        children
    }
    fn children_by_field_name(&self, field_name: &str) -> Vec<Self::Child> {
        match field_name {
            _ => vec![],
        }
    }
    fn children_by_field_id(&self, field_id: u16) -> Vec<Self::Child> {
        match field_id {
            _ => vec![],
        }
    }
}
impl<'db> std::hash::Hash for IfStatement<'db> {
    fn hash<H: std::hash::Hasher>(&self, state: &mut H) {
        self.id.hash(state);
    }
}
#[derive(Debug, Clone, Drive, Eq, PartialEq, salsa::Update)]
pub struct ReturnStatement<'db> {
    #[drive(skip)]
    start_byte: usize,
    #[drive(skip)]
    end_byte: usize,
    #[drive(skip)]
    _kind: std::string::String,
    #[drive(skip)]
    start_position: Point<'db>,
    #[drive(skip)]
    end_position: Point<'db>,
    #[drive(skip)]
    buffer: Arc<Bytes>,
    #[drive(skip)]
    kind_id: u16,
    #[drive(skip)]
    is_error: bool,
    #[drive(skip)]
    named: bool,
    #[drive(skip)]
    id: usize,
}
impl<'db> FromNode<'db> for ReturnStatement<'db> {
    fn from_node(
        db: &'db dyn salsa::Database,
        node: tree_sitter::Node,
        buffer: &Arc<Bytes>,
    ) -> Result<Self, ParseError> {
        let start_position = Point::from(db, node.start_position());
        let end_position = Point::from(db, node.end_position());
        Ok(Self {
            start_byte: node.start_byte(),
            end_byte: node.end_byte(),
            _kind: node.kind().to_string(),
            start_position: start_position,
            end_position: end_position,
            buffer: buffer.clone(),
            kind_id: node.kind_id(),
            is_error: node.is_error(),
            named: node.is_named(),
            id: node.id(),
        })
    }
}
impl<'db> CSTNode<'db> for ReturnStatement<'db> {
    fn kind(&self) -> &str {
        &self._kind
    }
    fn start_byte(&self) -> usize {
        self.start_byte
    }
    fn end_byte(&self) -> usize {
        self.end_byte
    }
    fn start_position(&self) -> Point<'db> {
        self.start_position
    }
    fn end_position(&self) -> Point<'db> {
        self.end_position
    }
    fn buffer(&self) -> &Bytes {
        &self.buffer
    }
    fn kind_id(&self) -> u16 {
        self.kind_id
    }
    fn is_error(&self) -> bool {
        self.is_error
    }
    fn is_named(&self) -> bool {
        self.named
    }
    fn id(&self) -> usize {
        self.id
    }
}
impl<'db1> HasChildren<'db1> for ReturnStatement<'db1> {
    type Child = Self;
    fn children(&self) -> Vec<Self::Child> {
        vec![]
    }
    fn children_by_field_name(&self, field_name: &str) -> Vec<Self::Child> {
        match field_name {
            _ => vec![],
        }
    }
    fn children_by_field_id(&self, field_id: u16) -> Vec<Self::Child> {
        match field_id {
            _ => vec![],
        }
    }
}
impl<'db> std::hash::Hash for ReturnStatement<'db> {
    fn hash<H: std::hash::Hasher>(&self, state: &mut H) {
        self.id.hash(state);
    }
}
#[salsa::tracked]
pub struct Parsed<'db> {
    #[tracked]
    #[return_ref]
    pub program: Option<Program<'db>>,
}
pub fn parse_program_raw(
    db: &dyn salsa::Database,
    input: codegen_sdk_cst::Input,
) -> Option<Program<'_>> {
    let buffer = Bytes::from(input.content(db).as_bytes().to_vec());
    let tree = codegen_sdk_common::language::language::Language
        .parse_tree_sitter(&input.content(db));
    match tree {
        Ok(tree) => {
            if tree.root_node().has_error() {
                ParseError::SyntaxError.report(db);
                None
            } else {
                let buffer = Arc::new(buffer);
                Program::from_node(db, tree.root_node(), &buffer)
                    .map_or_else(
                        |e| {
                            e.report(db);
                            None
                        },
                        |program| { Some(program) },
                    )
            }
        }
        Err(e) => {
            e.report(db);
            None
        }
    }
}
#[salsa::tracked]
pub fn parse_program(
    db: &dyn salsa::Database,
    input: codegen_sdk_cst::Input,
) -> Parsed<'_> {
    Parsed::new(db, parse_program_raw(db, input))
}
pub struct Language;
impl CSTLanguage for Language {
    type Program<'db> = Program<'db>;
    fn language() -> &'static Language {
        &codegen_sdk_common::language::language::Language
    }
    fn parse<'db>(
        db: &'db dyn salsa::Database,
        content: std::string::String,
    ) -> &'db Option<Self::Program<'db>> {
        let input = codegen_sdk_cst::Input::new(db, content);
        return parse_program(db, input).program(db);
    }
}
